File: src/main/java/no/novari/flyt/archive/gateway/dispatch/sak/CaseDispatchService.java
package no.novari.flyt.archive.gateway.dispatch.sak;

import io.netty.handler.timeout.ReadTimeoutException;
import lombok.extern.slf4j.Slf4j;
import no.fint.model.felles.kompleksedatatyper.Identifikator;
import no.fint.model.resource.arkiv.noark.SakResource;
import no.novari.flyt.archive.gateway.dispatch.mapping.SakMappingService;
import no.novari.flyt.archive.gateway.dispatch.model.instance.ArchiveInstance;
import no.novari.flyt.archive.gateway.dispatch.model.instance.SakDto;
import no.novari.flyt.archive.gateway.dispatch.sak.result.CaseDispatchResult;
import no.novari.flyt.archive.gateway.dispatch.sak.result.CaseSearchResult;
import no.novari.flyt.archive.gateway.dispatch.web.CreatedLocationPollTimeoutException;
import no.novari.flyt.archive.gateway.dispatch.web.FintArchiveDispatchClient;
import no.novari.flyt.archive.gateway.resource.web.CaseSearchParametersService;
import no.novari.flyt.archive.gateway.resource.web.FintArchiveResourceClient;
import no.novari.flyt.archive.gateway.resource.web.exceptions.KlasseOrderOutOfBoundsException;
import no.novari.flyt.archive.gateway.resource.web.exceptions.SearchKlasseOrderNotFoundInCaseException;
import org.springframework.stereotype.Service;
import org.springframework.web.reactive.function.client.WebClientResponseException;
import reactor.core.publisher.Mono;

@Service
@Slf4j
public class CaseDispatchService {
    private final SakMappingService sakMappingService;
    private final CaseSearchParametersService caseSearchParametersService;
    private final FintArchiveDispatchClient fintArchiveDispatchClient;
    private final FintArchiveResourceClient fintArchiveResourceClient;

    public CaseDispatchService(
            SakMappingService sakMappingService,
            CaseSearchParametersService caseSearchParametersService,
            FintArchiveDispatchClient fintArchiveDispatchClient, FintArchiveResourceClient fintArchiveResourceClient
    ) {
        this.sakMappingService = sakMappingService;
        this.caseSearchParametersService = caseSearchParametersService;
        this.fintArchiveDispatchClient = fintArchiveDispatchClient;
        this.fintArchiveResourceClient = fintArchiveResourceClient;
    }

    public Mono<CaseDispatchResult> dispatch(SakDto sakDto) {
        log.info("Dispatching case");
        SakResource sakResource = sakMappingService.toSakResource(sakDto);
        return fintArchiveDispatchClient.postCase(sakResource)
                .map(sr -> CaseDispatchResult.accepted(sr.getMappeId().getIdentifikatorverdi()))
                .doOnNext(caseDispatchResult -> log.info(
                        "Successfully posted case with archive case id = {}", caseDispatchResult.getArchiveCaseId()
                ))
                .onErrorResume(WebClientResponseException.class, e -> {
                            log.info("Post request for case was declined with message='{}'", e.getResponseBodyAsString());
                            return Mono.just(CaseDispatchResult.declined(e.getResponseBodyAsString()));
                        }
                ).onErrorResume(
                        e -> e instanceof ReadTimeoutException || e instanceof CreatedLocationPollTimeoutException,
                        e -> {
                            log.error("Case dispatch timed out");
                            return Mono.just(CaseDispatchResult.timedOut());
                        }
                )
                .onErrorResume(e -> {
                    log.error("Failed to post case", e);
                    return Mono.just(CaseDispatchResult.failed());
                }).doOnNext(result -> log.info("Dispatch result: {}", result.toString()));
    }

    public Mono<CaseSearchResult> findCasesBySearch(ArchiveInstance archiveInstance) {
        log.info("Searching for cases");
        try {
            String caseFilter = caseSearchParametersService.createFilterQueryParamValue(
                    archiveInstance.getNewCase(),
                    archiveInstance.getCaseSearchParameters()
            );
            log.debug("Generated case filter: {}", caseFilter);

            return fintArchiveResourceClient.findCasesWithFilter(caseFilter)
                    .map(sakResources -> sakResources.stream()
                            .map(SakResource::getMappeId)
                            .map(Identifikator::getIdentifikatorverdi)
                            .toList()
                    )
                    .map(CaseSearchResult::accepted)
                    .onErrorResume(ReadTimeoutException.class, e -> {
                        log.error("Case search timed out");
                        return Mono.just(CaseSearchResult.timedOut());
                    })
                    .onErrorResume(Exception.class, e -> {
                        log.error("Case search failed", e);
                        return Mono.just(CaseSearchResult.failed());
                    }).doOnNext(result -> log.info("Search result: {}", result.toString()));
        } catch (SearchKlasseOrderNotFoundInCaseException | KlasseOrderOutOfBoundsException e) {
            log.error("Case search failed", e);
            return Mono.just(CaseSearchResult.declined(e.getMessage()));
        } catch (Exception e) {
            log.error("Case search failed", e);
            return Mono.just(CaseSearchResult.failed());
        }

    }

}

File: src/main/kotlin/no/novari/flyt/archive/gateway/dispatch/sak/CaseDispatchService.kt
package no.novari.flyt.archive.gateway.dispatch.sak

import io.github.oshai.kotlinlogging.KotlinLogging
import no.novari.fint.model.felles.kompleksedatatyper.Identifikator
import no.novari.fint.model.resource.arkiv.noark.SakResource
import no.novari.flyt.archive.gateway.dispatch.isReadTimeout
import no.novari.flyt.archive.gateway.dispatch.mapping.SakMappingService
import no.novari.flyt.archive.gateway.dispatch.model.instance.ArchiveInstance
import no.novari.flyt.archive.gateway.dispatch.model.instance.SakDto
import no.novari.flyt.archive.gateway.dispatch.sak.result.CaseDispatchResult
import no.novari.flyt.archive.gateway.dispatch.sak.result.CaseSearchResult
import no.novari.flyt.archive.gateway.dispatch.web.CreatedLocationPollTimeoutException
import no.novari.flyt.archive.gateway.dispatch.web.FintArchiveDispatchClient
import no.novari.flyt.archive.gateway.resource.web.CaseSearchParametersService
import no.novari.flyt.archive.gateway.resource.web.FintArchiveResourceClient
import no.novari.flyt.archive.gateway.resource.web.exceptions.KlasseOrderOutOfBoundsException
import no.novari.flyt.archive.gateway.resource.web.exceptions.SearchKlasseOrderNotFoundInCaseException
import org.springframework.stereotype.Service
import org.springframework.web.client.RestClientResponseException

@Service
class CaseDispatchService(
    private val sakMappingService: SakMappingService,
    private val caseSearchParametersService: CaseSearchParametersService,
    private val fintArchiveDispatchClient: FintArchiveDispatchClient,
    private val fintArchiveResourceClient: FintArchiveResourceClient,
) {
    fun dispatch(sakDto: SakDto): CaseDispatchResult {
        log.info { "Dispatching case" }
        val sakResource: SakResource = sakMappingService.toSakResource(sakDto)

        val result =
            try {
                val resultSak = fintArchiveDispatchClient.postCase(sakResource)
                val archiveCaseId = resultSak.mappeId.identifikatorverdi
                log.atInfo {
                    message = "Successfully posted case with archive case id = {}"
                    arguments = arrayOf(archiveCaseId)
                }
                CaseDispatchResult.accepted(archiveCaseId)
            } catch (error: RestClientResponseException) {
                log.atInfo {
                    message = "Post request for case was declined with message='{}'"
                    arguments = arrayOf(error.responseBodyAsString)
                }
                CaseDispatchResult.declined(error.responseBodyAsString)
            } catch (error: CreatedLocationPollTimeoutException) {
                log.atError {
                    message = "Case dispatch timed out"
                    cause = error
                }
                CaseDispatchResult.timedOut()
            } catch (error: Throwable) {
                if (isReadTimeout(error)) {
                    log.atError {
                        message = "Case dispatch timed out"
                        cause = error
                    }
                    CaseDispatchResult.timedOut()
                } else {
                    log.atError {
                        message = "Failed to post case"
                        cause = error
                    }
                    CaseDispatchResult.failed()
                }
            }
        log.atInfo {
            message = "Dispatch result: {}"
            arguments = arrayOf(result)
        }
        return result
    }

    fun findCasesBySearch(archiveInstance: ArchiveInstance): CaseSearchResult {
        log.info { "Searching for cases" }

        val caseFilter =
            try {
                val newCase = requireNotNull(archiveInstance.newCase)
                val caseSearchParameters = requireNotNull(archiveInstance.caseSearchParameters)
                caseSearchParametersService.createFilterQueryParamValue(newCase, caseSearchParameters)
            } catch (error: SearchKlasseOrderNotFoundInCaseException) {
                log.atError {
                    message = "Case search failed"
                    cause = error
                }
                return CaseSearchResult.declined(error.message.orEmpty())
            } catch (error: KlasseOrderOutOfBoundsException) {
                log.atError {
                    message = "Case search failed"
                    cause = error
                }
                return CaseSearchResult.declined(error.message.orEmpty())
            } catch (error: Exception) {
                log.atError {
                    message = "Case search failed"
                    cause = error
                }
                return CaseSearchResult.failed()
            }
        log.atDebug {
            message = "Generated case filter: {}"
            arguments = arrayOf(caseFilter)
        }

        val result =
            try {
                val sakResources = fintArchiveResourceClient.findCasesWithFilter(caseFilter)
                val ids =
                    sakResources
                        .map(SakResource::getMappeId)
                        .map(Identifikator::getIdentifikatorverdi)
                CaseSearchResult.accepted(ids)
            } catch (error: Throwable) {
                if (isReadTimeout(error)) {
                    log.atError {
                        message = "Case search timed out"
                        cause = error
                    }
                    CaseSearchResult.timedOut()
                } else {
                    log.atError {
                        message = "Case search failed"
                        cause = error
                    }
                    CaseSearchResult.failed()
                }
            }
        log.atInfo {
            message = "Search result: {}"
            arguments = arrayOf(result)
        }
        return result
    }

    companion object {
        private val log = KotlinLogging.logger {}
    }
}
